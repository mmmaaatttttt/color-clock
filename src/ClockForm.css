.error .time,
.error p {
  color: red;
}

.error input.time {
  border: 1px solid red;
  border-radius: 2px;
}