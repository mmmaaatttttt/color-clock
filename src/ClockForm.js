import React from "react";

function ClockForm(props) {
  return <div>ClockForm</div>;
}

export default ClockForm;
