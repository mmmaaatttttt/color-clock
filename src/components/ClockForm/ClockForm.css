.ClockForm {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  flex-direction: column;
}

.ClockForm--form {
  width: 100%;
  margin: 0 auto;
  padding: 0 20px;
}

@media (min-width: 80rem) {
  .ClockForm--form {
    width: 50%;
  }

  .ClockForm {
    align-items: center;
  }
}

.ClockForm--form input[type="time"] {
  -webkit-appearance: none;
  background-color: transparent;
  border: 0.1rem solid #d1d1d1;
  border-radius: 0.4rem;
  box-shadow: none;
  box-sizing: inherit;
  height: 3.8rem;
  padding: 0.6rem 1rem 0.7rem;
  width: 100%;
}

.ClockForm--button.button {
  background-color: #2196f3;
  border-color: #2196f3;
  width: 100%;
  height: 5rem;
  margin-top: 1.5rem;
}

.ClockForm--form input:focus {
  border-color: #2196f3;
}

.ClockForm--button.button.hidden {
  pointer-events: none;
  opacity: 0;
}
