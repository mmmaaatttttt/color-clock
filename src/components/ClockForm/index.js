import React, { useState, useEffect, useMemo } from "react";
import { useHistory } from "react-router-dom";
import FormContext from "../../contexts/FormContext";
import ClockFormRow from "../ClockFormRow";
import Title from "../Title";
import { encodeToUrl } from "../../utils/routes";
import { DEFAULT_COLORS } from "../../data/colors";
import "./ClockForm.css";

/**
 * Form component for creating a new color clock.
 *
 * Form validates whether the times are in strictly ascending order,
 * and all three times must be provided in order to submit.
 * Upon successful form submit, the user is redirected to the
 * show page for the newly created clock.
 *
 * Props: none
 * State:
 *   - formData: {
 *       startTime: String,
 *       warningTime: String,
 *       endTime: String,
 *       startColor: String,
 *       warningColor: String,
 *       endColor: String
 *     }
 *   - errors: {
 *       warningTime: String,
 *       endTime: String,
 *     }
 *
 */
function ClockForm() {
  const history = useHistory();

  const [formData, setFormData] = useState({
    startTime: "",
    warningTime: "",
    endTime: "",
    startColor: DEFAULT_COLORS[0],
    warningColor: DEFAULT_COLORS[1],
    endColor: DEFAULT_COLORS[2]
  });

  const [errors, setErrors] = useState({
    warningTime: "",
    endTime: ""
  });

  const colors = useMemo(
    () => [formData.startColor, formData.warningColor, formData.endColor],
    [formData.startColor, formData.warningColor, formData.endColor]
  );

  useEffect(() => {
    const { startTime, warningTime, endTime } = formData;
    const isBefore = (time1, time2) => time1 && time2 && time1 <= time2;

    let warningError = "";
    let endError = "";
    if (isBefore(endTime, startTime)) {
      endError = "End time must come after start time.";
    }
    if (isBefore(warningTime, startTime) || isBefore(endTime, warningTime)) {
      warningError = "Warning time must be between start and end time.";
    }
    setErrors({ warningTime: warningError, endTime: endError });
  }, [formData]);

  const handleChange = e => {
    const { name, value } = e.target;
    setFormData(oldData => ({ ...oldData, [name]: value }));
  };

  const handleSubmit = e => {
    e.preventDefault();
    const noErrors = Object.values(errors).every(err => err === "");
    if (noErrors) {
      const times = [
        formData.startTime,
        formData.warningTime,
        formData.endTime
      ];
      const clockId = encodeToUrl({ colors, times });
      history.push(`/clocks/${clockId}`);
    }
  };

  const timesFilledOut =
    formData.startTime && formData.warningTime && formData.endTime;

  return (
    <FormContext.Provider value={{ handleChange, formData, errors }}>
      <div className="ClockForm">
        <Title colors={colors} />
        <form onSubmit={handleSubmit} className="ClockForm--form">
          <ClockFormRow
            prefix="start"
            title="When do you want the clock to start?"
          />
          <ClockFormRow prefix="warning" title="When do you want to warn?" />
          <ClockFormRow prefix="end" title="When is time up?" />
          <input
            type="submit"
            className={`ClockForm--button button ${
              timesFilledOut ? "" : "hidden"
            }`}
            value="Create clock!"
          />
        </form>
      </div>
    </FormContext.Provider>
  );
}

export default React.memo(ClockForm);
