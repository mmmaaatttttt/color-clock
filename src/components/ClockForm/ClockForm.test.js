import React from "react";
import { render } from "@testing-library/react";
import ClockForm from "./";

describe("basic ClockForm tests", function () {
  it("renders without crashing", function () {
    render(<ClockForm />);
  });

  it("matches snapshot", function () {
    const { asFragment } = render(<ClockForm />);
    expect(asFragment()).toMatchSnapshot();
  });
});
