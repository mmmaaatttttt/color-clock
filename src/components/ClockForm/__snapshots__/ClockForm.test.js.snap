// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic ClockForm tests matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ClockForm"
  >
    <form
      class="ClockForm--form"
    >
      <div>
        <h4
          class="ClockFormRow--heading"
        >
          When do you want the clock to start?
        </h4>
        <div
          class="ClockFormRow row "
        >
          <div
            class="column time"
          >
            <label
              for="startTime"
            >
              Start Time
            </label>
            <input
              id="startTime"
              name="startTime"
              type="time"
              value=""
            />
          </div>
          <div
            class="column"
          >
            <label
              for="startColor"
            >
              Start Color
            </label>
            <input
              id="startColor"
              name="startColor"
              type="color"
              value="#00ff00"
            />
          </div>
        </div>
      </div>
      <div>
        <h4
          class="ClockFormRow--heading"
        >
          When do you want to warn?
        </h4>
        <div
          class="ClockFormRow row "
        >
          <div
            class="column time"
          >
            <label
              for="warningTime"
            >
              Warning Time
            </label>
            <input
              id="warningTime"
              name="warningTime"
              type="time"
              value=""
            />
          </div>
          <div
            class="column"
          >
            <label
              for="warningColor"
            >
              Warning Color
            </label>
            <input
              id="warningColor"
              name="warningColor"
              type="color"
              value="#ffff00"
            />
          </div>
        </div>
      </div>
      <div>
        <h4
          class="ClockFormRow--heading"
        >
          When is time up?
        </h4>
        <div
          class="ClockFormRow row "
        >
          <div
            class="column time"
          >
            <label
              for="endTime"
            >
              End Time
            </label>
            <input
              id="endTime"
              name="endTime"
              type="time"
              value=""
            />
          </div>
          <div
            class="column"
          >
            <label
              for="endColor"
            >
              End Color
            </label>
            <input
              id="endColor"
              name="endColor"
              type="color"
              value="#ff0000"
            />
          </div>
        </div>
      </div>
      <input
        class="ClockForm--button button hidden"
        type="submit"
        value="Create clock!"
      />
    </form>
  </div>
</DocumentFragment>
`;
